# Assignments

## Assignment 1

### DataFrame Formatting

In [1]:
# Set up shell for libraries, packages, and data

import pandas as pd
import numpy as np

metadata = pd.read_csv('../data/meta_data.csv')
df = pd.read_csv('../data/transaction_data.csv')

In [2]:
metadata


feature                                        description
0          sku_number          Eight character stock keeping unit number
1      inventory_type  Types of inventory including finished good - w...
2       stocking_type  Stockign type policy including make to stock -...
3           lead_time                            Total lead time in days
4          unit_price                         Unit price US dollar value
5  manufacturing_site  Manufacturing site ID starting with country co...
6       division_code                Four character business division ID
7    transaction_date       Date of Transaction in Year-Month-Day format
8      order_quantity        Order quantity - rounded up for convenience

In [3]:
df

sku_numb3r inventory_type stocking_type  leadTime   unit_price  \
0        E38C9BDE             FG           MTO        28  1004.368931   
1        43BA1933             FG           MTO        28   994.515536   
2        1D2305AE             FG           MTO        21   200.206630   
3        E8B8CAB8             RM           MTO        28  1002.689160   
4        25287B58            WIP           MTO        28  1005.488658   
...           ...            ...           ...       ...          ...   
400228   985B2405             FG           MTS        28   994.202002   
400229   EEFAC643             FG           MTS        28  1013.021153   
400230   B4EF1CFC             FG           MTS        28    93.902638   
400231   D5EBA9EA             FG           MTO        28  1010.663326   
400232   A4449403             FG           NaN        28  1014.515948   

       manufacturing_site division_code transaction_date  Order-Quantity  
0                     NaN          A9A4       2023-01-01              61  
1                   CA-1C          B9F3       2023-01-01             105  
2                     NaN          2BFB       2023-01-01               3  
3                   CA-1C           NaN       2023-01-01             114  
4                   US-E7          8F01       2023-01-01             104  
...                   ...           ...              ...             ...  
400228              PL-BD           NaN       2024-12-30             151  
400229              CA-1C          8CB1       2024-12-30             120  
400230                NaN          4879       2024-12-30              16  
400231                NaN          0305       2024-12-30             117  
400232              JP-B0          2BFB       2024-12-30              97  

[400233 rows x 9 columns]

In [4]:
# Dataframe Formatting ~ Changing column names to follow the same naming convention

# Renaming columns for "transaction" dataframe

df.rename(columns={'sku_numb3r': 'sku_number','leadTime': 'lead_time','Division Code':'division_code','Order-Quantity':'order_quantity'}, inplace=True)

# view df column names as list:

print(df.columns.tolist())

['sku_number', 'inventory_type', 'stocking_type', 'lead_time', 'unit_price', 'manufacturing_site', 'division_code', 'transaction_date', 'order_quantity']


### Data Inspection

In [5]:
# inspect the numeric features of the transaction dataframe
df.describe()

lead_time     unit_price  order_quantity
count  400233.000000  400233.000000   400233.000000
mean       25.752612     798.225116       79.948075
std         6.474507     374.181148       41.917447
min        14.000000    -994.354714     -119.000000
25%        28.000000     980.888516       63.000000
50%        28.000000     995.185255       93.000000
75%        28.000000    1003.639787      109.000000
max      2800.000000    1026.646918      191.000000

 **Data Quality Issues (Oddities & Anomolies):**

* Missing Values- `manufacturing_site`, `division_code`, and `stocking_type` contain several missing values (NaN & NA).

* Outliers in `unit_price` - column contains at least one observation with a negative value; most values in the data set average around ~$800, there is at least one obervation with unit price of -$994.

* Outliers in `order_quantity` - most quantites are in double-digits, at least one observation has a low order number (3).


**Changes and Recommendations:**

* Remove NaN values and "NA" strings to preserve data integrity.
* Apply filter to `unit_price` by removing negative values and filter `order_quantity` to only reflect observations with a value of 10 or higher to remove outliers and avoid skewing. 



**Summary of Data Inspection:**

The summary statistics reveal a few important patterns and some clear anomalies. For example, `lead_time` appears highly consistent, with the 25th, 50th, and 75th percentiles all at 28 days which suggests a standardized process. Furhtermore, `unit_price` is tightly clustered between approximately $981 and $1004, which indicates price stability across most of the records. However, the minimum value of –$994 is possibly invalid and may indicate a misrecorded transaction or a potential return/refund. Additionnally, `order_quantity` shows a reasonable distribution centered around 80 units, however, the presence of a –119 minimum suggests a possible data issue, possibly related to returns or data entry errors. Lastly, the data needs to be inspected to ensure there are no duplicate entries. Overall, while the data appears generally well-structured but the extreme outliers, missing, and negative values should be flagged and addressed prior to proceeding with deeper analysis.


### Handling `NaN` Values

In [ ]:
# drop na's for sku_number

df = df.dropna(subset=['sku_number'])

# Drop rows where sku_numner column contains the string "NA"
df = df[~df[['sku_number']].isin(['NA']).any(axis=1)]

# handling negative unit_price values by converting to absolute values & filtering out negative order_quantity values

df['unit_price'] = df['unit_price'].abs()
df = df[df['order_quantity'] >= 0]

df

sku_number inventory_type stocking_type  lead_time   unit_price  \
0        E38C9BDE             FG           MTO         28  1004.368931   
1        43BA1933             FG           MTO         28   994.515536   
2        1D2305AE             FG           MTO         21   200.206630   
3        E8B8CAB8             RM           MTO         28  1002.689160   
4        25287B58            WIP           MTO         28  1005.488658   
...           ...            ...           ...        ...          ...   
400228   985B2405             FG           MTS         28   994.202002   
400229   EEFAC643             FG           MTS         28  1013.021153   
400230   B4EF1CFC             FG           MTS         28    93.902638   
400231   D5EBA9EA             FG           MTO         28  1010.663326   
400232   A4449403             FG           NaN         28  1014.515948   

       manufacturing_site division_code transaction_date  order_quantity  \
0                     NaN          A9A4       2023-01-01              61   
1                   CA-1C          B9F3       2023-01-01             105   
2                     NaN          2BFB       2023-01-01               3   
3                   CA-1C           NaN       2023-01-01             114   
4                   US-E7          8F01       2023-01-01             104   
...                   ...           ...              ...             ...   
400228              PL-BD           NaN       2024-12-30             151   
400229              CA-1C          8CB1       2024-12-30             120   
400230                NaN          4879       2024-12-30              16   
400231                NaN          0305       2024-12-30             117   
400232              JP-B0          2BFB       2024-12-30              97   

        total_sales_value  
0            61266.504814  
1           104424.131232  
2              600.619891  
3           114306.564292  
4           104570.820446  
...                   ...  
400228      150124.502251  
400229      121562.538350  
400230        1502.442208  
400231      118247.609137  
400232       98408.046997  

[379823 rows x 10 columns]

`NaN` Handling Notes:

- Removed incomplete records by dropping rows with missing (NaN) values or placeholder "NA" strings in `sku_number` column to ensure data integrity, reliability, and accuracy for categorization and future analyses.

- Filtered outliers/ unusually high or low values in `unit_price` and `order_quantity` to avoid misleading results.



 ### Useful Information

In [17]:
# Determine unique sku_number count

sku_count = df['sku_number'].nunique()
print(f"Number of unique values in 'sku_number' column: {sku_count}")

Number of unique values in 'sku_number' column: 473


In [9]:
# Determine unique manufacturing_site count

manufacturing_count = df['manufacturing_site'].nunique()
print(f"Number of unique values in 'manufacturing_site' column: {manufacturing_count}")

Number of unique values in 'manufacturing_site' column: 15


In [10]:
# Determine division_code count

division_count = df['division_code'].nunique()
print(f"Number of unique values in 'division_code' column: {division_count}")

Number of unique values in 'division_code' column: 66


In [25]:
# Top 10 transactions by order_quantity

top_orders = df.nlargest(10, 'order_quantity')
top_orders

sku_number inventory_type stocking_type  lead_time   unit_price  \
10594    8EF136C1             FG           ATO         28  1012.003538   
171719   17A73316             FG           MTS         28   992.194096   
388055   6C5EA2F2             RM           MTO         28   989.771998   
69516    2179D986             FG           MTS         28  1008.453792   
367181   A4449403             FG           NaN         28  1014.515948   
72063    2D2C2A10            WIP           MTS         28   995.540635   
213431   3204B87F             RM           MTO         28  1004.316620   
42620    CD903AB0             RM           MTO         28  1004.122690   
195971   1BE106F5             FG           MTS         28   994.092509   
214314   E5CDF84B             FG           MTS         28  1000.613975   

       manufacturing_site division_code transaction_date  order_quantity  
10594                 NaN          0305       2023-01-20             191  
171719                NaN          6777       2023-11-10             191  
388055              PL-35          A93E       2024-12-08             188  
69516                 NaN          8F01       2023-05-07             186  
367181              JP-E9          F076       2024-10-31             186  
72063               JP-B0          3AB7       2023-05-12             185  
213431              CA-1C          B9F3       2024-01-24             183  
42620               CA-1C          666E       2023-03-19             182  
195971              US-E7          1C83       2023-12-24             182  
214314              JP-E9          0305       2024-01-26             182

In [26]:
# Bottom 10 transactions by order_quantity

bottom_orders = df.nsmallest(10, 'order_quantity')
bottom_orders

sku_number inventory_type stocking_type  lead_time  unit_price  \
57     595261EC             RM           MTO         14   42.659739   
117    02E5FD89             FG           MTS         21  208.946153   
148    A250411D             FG           MTO         21  192.205044   
179    85F64A1F             RM           MTS         28  100.545087   
449    BD670FE2            NaN           MTS         14   45.702916   
525    71395CC1             FG           MTO         21  192.032849   
680    82CC5208             FG           ATO         14   43.620770   
1069   8EF136C1             RM           MTS         14   47.511957   
1547   C7906166            WIP           MTS         14   42.738155   
1571   6BA32785            WIP           MTS         14   40.654232   

     manufacturing_site division_code transaction_date  order_quantity  
57                  NaN          6777       2023-01-01               0  
117               JP-E9           NaN       2023-01-01               0  
148               CA-1C          597C       2023-01-01               0  
179               CA-1C          65F7       2023-01-01               0  
449               CA-15           NaN       2023-01-01               0  
525                 NaN          8CB1       2023-01-01               0  
680               PL-35          666E       2023-01-02               0  
1069              JP-E9          739B       2023-01-02               0  
1547                NaN          739B       2023-01-03               0  
1571                NaN          1C83       2023-01-03               0

In [27]:
# Top 10 transactions by total_sales_value

# create total_sales_value column and calculate transactions

df['total_sales_value'] = (df['unit_price'] * df['order_quantity'])

# Calculate and display top 10 sales transactions

top_sales = df.sort_values(by='total_sales_value', ascending=False).head(10)
top_sales.style.format({'total_sales_value': '${:,.2f}'})

In [14]:
# Bottom 10 transactions by total_sales_value

bottom_sales = df.sort_values(by='total_sales_value', ascending=True).head(10).round(2)
bottom_sales.style.format({'total_sales_value': '${:,.2f}'})

### References



1.	GeeksforGeeks – Working with Missing Data in Pandas
GeeksforGeeks. “Working with Missing Data in Pandas.” GeeksforGeeks, 28 July 2025, https://www.geeksforgeeks.org/data-analysis/working-with-missing-data-in-pandas/.

2.	GeeksforGeeks – Pandas: How to Use dropna() with Specific Columns
GeeksforGeeks. “Pandas DataFrame.dropna() Method.” GeeksforGeeks, 25 June 2025, https://www.geeksforgeeks.org/python/python-pandas-dataframe-dropna/.

3.	GeeksforGeeks – How to Find Duplicates in Pandas DataFrame (With Examples)
GeeksforGeeks. “Find Duplicate Rows in a Dataframe Based on All or Selected Columns.” GeeksforGeeks, 4 Dec. 2023, https://www.geeksforgeeks.org/python/find-duplicate-rows-in-a-dataframe-based-on-all-or-selected-columns/.

4.	GeeksforGeeks – Filter Pandas Dataframe with Multiple Conditions
GeeksforGeeks. “Filter Pandas Dataframe with Multiple Conditions.” GeeksforGeeks, 23 July 2025, https://www.geeksforgeeks.org/python/filter-pandas-dataframe-with-multiple-conditions/.

5.	GeeksforGeeks – Adding New Column to Existing DataFrame in Pandas
GeeksforGeeks. “Adding New Column to Existing DataFrame in Pandas.” GeeksforGeeks, 11 July 2025, https://www.geeksforgeeks.org/pandas/adding-new-column-to-existing-dataframe-in-pandas/.

6. GeeksforGeeks. Get n-Smallest Values from a Particular Column in Pandas DataFrame. GeeksforGeeks, 11 July 2025, https://www.geeksforgeeks.org/python/get-n-smallest-values-from-a-particular-column-in-pandas-dataframe/.

7.	Pandas Documentation – Working with Missing Data
Pandas Development Team. “Working with Missing Data.” Pandas 2.3.3 Documentation, pandas.pydata.org, https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html.

8.	Stackoverflow. (2012). "Renaming column names in Pandas". Retrieved October 1st, 2025. https://stackoverflow.com/questions/11346283/renaming-column-names-in-pandas

9.	Stackoverflow. (2012). "How to show all columns' names on a large pandas dataframe?Retrieved September 30th, 2025. https://stackoverflow.com/questions/11346283/renaming-column-names-in-pandas




## Assignment 2

In [18]:
# Set up shell for data and packages

import pandas as pd
from scipy.stats import norm
import numpy as np

# Load dataset
metadata = pd.read_csv('../data/meta_data.csv')
df = pd.read_csv('../data/transaction_data.csv')

In [19]:
# Dataframe Formatting 

# Renaming columns for "transaction" dataframe

df.rename(columns={'sku_numb3r': 'sku_number','leadTime': 'lead_time','Division Code':'division_code','Order-Quantity':'order_quantity'}, inplace=True)

# drop na's for sku_number

df = df.dropna(subset=['sku_number'])

# handling negative unit_price values by converting to absolute values & filtering out negative order_quantity values

df['unit_price'] = df['unit_price'].abs()
df = df[df['order_quantity'] >= 0]

df

sku_number inventory_type stocking_type  lead_time   unit_price  \
0        E38C9BDE             FG           MTO         28  1004.368931   
1        43BA1933             FG           MTO         28   994.515536   
2        1D2305AE             FG           MTO         21   200.206630   
3        E8B8CAB8             RM           MTO         28  1002.689160   
4        25287B58            WIP           MTO         28  1005.488658   
...           ...            ...           ...        ...          ...   
400228   985B2405             FG           MTS         28   994.202002   
400229   EEFAC643             FG           MTS         28  1013.021153   
400230   B4EF1CFC             FG           MTS         28    93.902638   
400231   D5EBA9EA             FG           MTO         28  1010.663326   
400232   A4449403             FG           NaN         28  1014.515948   

       manufacturing_site division_code transaction_date  order_quantity  
0                     NaN          A9A4       2023-01-01              61  
1                   CA-1C          B9F3       2023-01-01             105  
2                     NaN          2BFB       2023-01-01               3  
3                   CA-1C           NaN       2023-01-01             114  
4                   US-E7          8F01       2023-01-01             104  
...                   ...           ...              ...             ...  
400228              PL-BD           NaN       2024-12-30             151  
400229              CA-1C          8CB1       2024-12-30             120  
400230                NaN          4879       2024-12-30              16  
400231                NaN          0305       2024-12-30             117  
400232              JP-B0          2BFB       2024-12-30              97  

[379823 rows x 9 columns]

### Required Transformation

In [20]:
# filter for appropriate SKUs

filtered_df = df[(df['inventory_type'] == 'FG') & (df['stocking_type'] == 'MTS')]
grouped_df = filtered_df.groupby('sku_number')
grouped_df.head()



sku_number inventory_type stocking_type  lead_time   unit_price  \
6        9754C645             FG           MTS         28   991.679520   
10       DEF15D18             FG           MTS         28  1003.918849   
11       00749B73             FG           MTS         28  1009.936435   
14       A250411D             FG           MTS         28  1005.801932   
16       9E9B370E             FG           MTS         28   987.252629   
...           ...            ...           ...        ...          ...   
87302    5FD68AE1             FG           MTS         14    57.532735   
88746    5D8F72C0             FG           MTS         14    53.857587   
106193   296D2C2C             FG           MTS         14    56.476529   
124203   90EA57FF             FG           MTS         14    55.880150   
126525   90EA57FF             FG           MTS         14    55.880150   

       manufacturing_site division_code transaction_date  order_quantity  
6                   JP-E9          0305       2023-01-01              80  
10                  JP-E9          8F01       2023-01-01              65  
11                    NaN          8F01       2023-01-01             103  
14                  US-6D          0305       2023-01-01             100  
16                    NaN          B9F3       2023-01-01              85  
...                   ...           ...              ...             ...  
87302                 NaN          666E       2023-06-09               7  
88746                 NaN           NaN       2023-06-11              14  
106193              JP-E9          9ED0       2023-07-13               2  
124203              JP-E9          739B       2023-08-15              14  
126525              CA-8C          FDED       2023-08-19               9  

[1465 rows x 9 columns]

In [21]:

# Aggregate order quantity stats per SKU group

aggregate_df = grouped_df.aggregate({'order_quantity': ['min', 'max', 'mean', 'median', 'var', 'std'],'lead_time': 'mean'})
aggregate_df.head()


order_quantity                                                 \
                      min  max        mean median         var        std   
sku_number                                                                 
0061F576                0   32   10.675676   10.0   70.891892   8.419732   
00749B73               33  166   99.657673   99.0  399.040519  19.975999   
02E5FD89                0   67   24.918919   21.0  267.993396  16.370504   
0303002D                0   27    9.439024    8.0   46.652439   6.830259   
0309C11F               39  173  100.998381  102.0  375.379252  19.374706   

           lead_time  
                mean  
sku_number            
0061F576        14.0  
00749B73        28.0  
02E5FD89        21.0  
0303002D        14.0  
0309C11F        28.0

### Safety Stock Calculation

In [23]:
# Calculate the safety stock for each SKU for the service level of 75%

# Compute the value at the 75th percentile
a75 = 0.75      
alpha_75 = norm.ppf(a75)

# calculate safety stock 
aggregate_df['safety_stock_75'] = (
    alpha_75 * aggregate_df[('order_quantity', 'std')] * np.sqrt(aggregate_df[('lead_time', 'mean')])
)


# Compute the value at the 90th percentile
a90 = 0.90
alpha_90 = norm.ppf(a90)

# calculate safety stock 

aggregate_df['safety_stock_90'] = (
    alpha_90 * aggregate_df[('order_quantity', 'std')] * np.sqrt((aggregate_df[('lead_time', 'mean')])))



# Compute the value at the 95th percentile
a95 = 0.95
alpha_95 = norm.ppf(a95)

# calculate safety stock 

aggregate_df['safety_stock_95'] = (
    alpha_95 * aggregate_df[('order_quantity', 'std')] * np.sqrt((aggregate_df[('lead_time', 'mean')])))

rounded_aggregate_df = aggregate_df.apply(np.ceil).astype(int)
rounded_aggregate_df



order_quantity                           lead_time safety_stock_75  \
                      min  max mean median  var std      mean                   
sku_number                                                                      
0061F576                0   32   11     10   71   9        14              22   
00749B73               33  166  100     99  400  20        28              72   
02E5FD89                0   67   25     21  268  17        21              51   
0303002D                0   27   10      8   47   7        14              18   
0309C11F               39  173  101    102  376  20        28              70   
...                   ...  ...  ...    ...  ...  ..       ...             ...   
FB2A918D                0   39   11     10   65   9        14              21   
FB818D5C                0  156   95     97  654  26        28              91   
FBC62BD7                0   28   11     11   54   8        14              19   
FE3FA206                0  179   97     99  700  27        28              94   
FF160ACE               41  159  101    101  393  20        28              71   

           safety_stock_90 safety_stock_95  
                                            
sku_number                                  
0061F576                41              52  
00749B73               136             174  
02E5FD89                97             124  
0303002D                33              43  
0309C11F               132             169  
...                    ...             ...  
FB2A918D                39              50  
FB818D5C               172             221  
FBC62BD7                36              46  
FE3FA206               178             228  
FF160ACE               135             173  

[293 rows x 10 columns]

NOTES: added the`.apply(np.ceil).astype(int)` function to round safety stock calculation to the nearest whole number because you cant have partial units of inventory, however this doesn't provide the most accurate measure of safety stock because some are rounded up and some are rounded down (depending on the decimal point)

### Safety Stock Distribution

In [24]:
# Determine which SKU has the largest safety stock at 95% service level
max_safety_stock_95 = aggregate_df['safety_stock_95'].idxmax()
print("SKU with the largest safety stock at the 95th percentile:", max_safety_stock_95)


# Determine which SKU has the smallest safety stock at 95% service level
min_safety_stock_95 = aggregate_df['safety_stock_95'].idxmin()
print("SKU with the smallest safety stock at the 95th percentile:", min_safety_stock_95)

# Calculate average safety stock across all SKUs at 95% service level
average_safety_stock_95 = aggregate_df['safety_stock_95'].mean().round().astype(int)
print("Average safety stock at the 95th percentile:", average_safety_stock_95)


SKU with the largest safety stock at the 95th percentile: BE575720
SKU with the smallest safety stock at the 95th percentile: 79536466
Average safety stock at the 95th percentile: 109


### References

* Bobbitt, Zach. “Pandas: How to Use dropna() with Specific Columns.” Statology, 13 Feb. 2023, https://www.statology.org/pandas-dropna-specific-column/.


* Kumar, Bijay. “Scipy Stats Zscore: Calculate and Use Z-Score.” *Python Guides*, 20 June 2025, https://pythonguides.com/scipy-stats-zscore/. Accessed 21 Oct. 2025.

* NumPy Developers. “numpy.round — NumPy v2.3 Manual.” *NumPy*, https://numpy.org/doc/stable/reference/generated/numpy.round.html. Accessed 21 Oct. 2025.

* pandas.DataFrame.aggregate — pandas 2.3.3 Documentation.” *pandas*, PyData, https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.aggregate.html.





